source/common/singleton/manager_impl.cc

```diff
@@ -0,0 +1,26 @@
+#include "common/singleton/manager_impl.h"
+
+#include "envoy/registry/registry.h"
+
+namespace Envoy {
+namespace Singleton {
+
+InstancePtr ManagerImpl::get(const std::string& name) {
+  verifyRegistration(name);
+  return singletons_[name].lock();
+}
+
+void ManagerImpl::set(const std::string& name, InstancePtr singleton) {
+  verifyRegistration(name);
+  singletons_[name] = singleton;
+}
+
+void ManagerImpl::verifyRegistration(const std::string& name) {
+  if (nullptr == Registry::FactoryRegistry<Registration>::getFactory(name)) {
```

Comment:
I think it probably needs to be more than an ASSERT since we want to verify the name is unique. I could make it a PANIC() with appropriate message. I just did exception since it's easier to test. No real preference either way (could also switch to an illegal arg exception which would be more likely to crash).